src/Microsoft.Extensions.Configuration.AzureAppConfiguration/ConfigurationClients/FailOverSupportedConfigurationClient.cs

```diff
@@ -0,0 +1,253 @@
+                    }
+                }
+            }
+
+            if (replicasTried == 0)
+            {
+                return await ExecuteWithFailOverPolicyAsync(funcToExecute, forceTryFailedReplicas: true, cancellationToken);
+            }
+
+            throw lastException;
+        }
+
+        private async Task<T> ExecuteWithFailOverPolicyAsync<T>(Func<ConfigurationClient, T> funcToExecute, bool forceTryFailedReplicas = false, CancellationToken cancellationToken = default)
+        {
+            var tasks = new List<Task>();
+            var attemptsCancellation = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
+            int replicasTried = 0;
+            Exception lastException = null;
+
+            for (int i = 0; i < _configurationClientAndStates.Count(); i++)
+            {
+                ConfigurationClient client = _configurationClientAndStates.ElementAt(i).Item1;
+                ConfigurationClientState clientState = _configurationClientAndStates.ElementAt(i).Item2;
+
+                if (forceTryFailedReplicas || clientState.IsAvailable())
+                {
+                    var parallelAwait = Task.Delay(this._parallelRetryInterval, attemptsCancellation.Token);
+                    Task<T> funcTask = Task.Run(() => funcToExecute(client), cancellationToken);
+                    tasks.Add(funcTask);
+                    tasks.Add(parallelAwait);
+                    replicasTried++;
+
+                    var completedTask = await Task.WhenAny(tasks);
+                    if (!completedTask.Equals(parallelAwait))
+                    {
+                        if (completedTask.Status == TaskStatus.RanToCompletion)
+                        {
+                            clientState.UpdateConfigurationStoreStatus(requestSuccessful: true);
+
+                            // Safe because the task is completed.
+                            return funcTask.Result;
+                        }
+                        else if (completedTask.Status == TaskStatus.Faulted)
+                        {
+                            tasks.Clear();
+
+                            if (IsRetryableException(completedTask.Exception))
+                            {
+                                lastException = completedTask.Exception;
+                                clientState.UpdateConfigurationStoreStatus(requestSuccessful: false);
+                            }
+                            else
+                            {
+                                throw completedTask.Exception;
+                            }
+                        }
+                    }
+                    else
+                    {
+                        tasks.Remove(parallelAwait);
+                    }
+                }
+            }
+
+            if (replicasTried == 0)
+            {
+                return await ExecuteWithFailOverPolicyAsync<T>(funcToExecute, forceTryFailedReplicas: true, cancellationToken);
```

Comment:
Recursion adds complexity. Can we construct the list of clients at the beginning and ensure nonempty to avoid recursion?

Reply:
We can do that, but that would mean for a given request, we won't have most up-to-date information about the replica's status, and we would be trying failed replicas.

i.e., A concurrent request could modify the status of a configuration client to `failed` but since we have a precomputed list of clients, we would still try the failed replica.

Reply:
With the new design, no need for recursion.